### PR TITLE
Read/Write iterators

### LINKED ISSUE #8: Remove Deref/DerefMut
Instead, have a custom iterator with an ability to spawn pointers on the way.
cc @vitvakatu

### PR DESCRIPTION
**Pull request by kvark**
Replaces #11
Fixes #8

The only (and important) downside of the PR is the introduction of `unsafe` code for `WriteItem`.

### CODE REVIEW

**Comment by vitvakatu** [comment]
Well, I like it more than my attempt :)
If it's final decision, you can merge it and I'll fix demo.

File: README.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # froggy
-[![Build Status](https://travis-ci.org/kvark/froggy.svg)](https://travis-ci.org/kvark/froggy)
+[![Build Status](https://travis-ci.org/kvark/froggy.svg?branch=master)](https://travis-ci.org/kvark/froggy)
 [![Crates.io](https://img.shields.io/crates/v/froggy.svg?maxAge=2592000)](https://crates.io/crates/froggy)
+[![Gitter](https://badges.gitter.im/kvark/froggy.svg)](https://gitter.im/almost-ecs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
```
Comment (torkleyy) [inline README.md]: A link/badge for docs might also be nice.

Comment (kvark) [inline README.md]: agreed, fixed now

File: src/lib.rs [hunk 2]
```diff
@@ -158,11 +174,35 @@ impl<T> Drop for Pointer<T> {
...
+impl<'a, T> Iterator for ReadIter<'a, T> {
+    type Item = ReadItem<'a, T>;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let id = self.index;
+            if id == self.lock.guard.data.len() {
+                return None
+            }
+            self.index += 1;
+            if !self.skip_lost || self.lock.guard.meta[id] != 0 {
```
Comment (torkleyy) [inline src/lib.rs]: Do you think the compiler can move the branch out of the loop or should we have `ReadIterAlive`?

Comment (kvark) [inline src/lib.rs]: I don't think this is a big performance concern at the moment. If not anything else, the hardware CPU branch predictor will remember to always go the right path

**Comment by kvark** [comment]
Ok, I believe that everyone had a chance to look at this:
  - the replacement of slice derefs with new iterators is a confirmed decision, so even if we decide to change it, we can do this incrementally
  - pointer advancement is new and experimental, but it doesn't introduce new concepts, so not concerning much
  - cloning during the scene graph traversal is unfortunate, we'll look into fixing it without stalling the rest of the changes

**Pull request merged**